### PR TITLE
[GUILIB] add override onclick to dynamic content lists

### PR DESCRIPTION
**Pull request by InuSasha**
<content list="videodb://recentlyaddedmovies/">
&nbsp;&nbsp;<onclick>PlayMedia("$INFO[ListItem.FileNameAndPath]")</onclick>
</content>

if no onclick is defined, the default behavior is used.
the <content>videodb://..</content> is still working, too

additional: ExecuteAction methods gets an optional item as parameter.
This is used for ListItem infolabel while parse action string.

### CODE REVIEW

**Comment by InuSasha** [comment]
@phil65 ping

**Comment by ronie** [comment]
great, very cool :-)

@mkortstiege for review

**Comment by phil65** [comment]
Thx first of all.
Is there any advantage by using the "list" attribute + nesting instead of the previous structure? It does not seem very consistent that way (onback and soon oninfo are just an additional xml element for the container)

**Comment by InuSasha** [comment]
a child node will work too, like:
```xml
<content>
     <list>videodb://...</list>
     <onclick>DoStuff<onclick>
</content>
```
the actual solution was easy to implement, for me.

**Comment by phil65** [comment]
I thaught about simply
```xml
<control type="list">
...
<onclick>some_builtin</onclick>
<content>...</content>
</control>
```

**Comment by InuSasha** [comment]
okay at this point, i haven't enough overview over the code of kodi.
Sorry :(

**Comment by InuSasha** [comment]
@phil65 i had look into to the PR8177 from mkortstiege and have adopted it.
The result you can see in https://github.com/InuSasha/xbmc/commit/3940dc7cf5393d56e26ccadf4e1de4562eefd5d3

@mkortstiege i think, it is your decision
if we use the new variation, tell me, i update this PR.

**Comment by mkortstiege** [comment]
@InuSasha thanks for working on this! The new variation looks better to me as it's more natural for skinners. @phil65, @ronie what do you guys think?

**Comment by ronie** [comment]
yep, i agree with @phil65 
so second variation has my preference as well

**Comment by InuSasha** [comment]
PR is switched to the second variation

File: xbmc/guilib/GUIBaseContainer.cpp [hunk 1]
```diff
@@ -752,7 +752,10 @@ bool CGUIBaseContainer::OnClick(int actionID)
     { // "select" action
       int selected = GetSelectedItem();
       if (selected >= 0 && selected < (int)m_items.size())
-        m_listProvider->OnClick(m_items[selected]);
```
Comment (mkortstiege) [inline xbmc/guilib/GUIBaseContainer.cpp]: Please add brackets here.

**Comment by InuSasha** [comment]
added

**Comment by mkortstiege** [comment]
Thanks. jenkins build this please.

**Comment by BigNoid** [comment]
Nice! Works as intended from a quick test run.
This doesn't override onclicks for lists without dynamic content, which is a good thing imo. But that makes the second variation a bit unclear. The first variation made clear directly that it was only for dynamic content.

**Comment by phil65** [comment]
Hmm true, if only dynamic content is affected then @BigNoid ´s point is valid. I do not really have a strong opinion on which one to use as long as this functionality makes it in. :)

**Comment by InuSasha** [comment]
The impact on static-list will small, because without an onclick element in the list-control, the default behavior is used.

**Comment by BigNoid** [comment]
I am okay with this implementation too, just pointing out that it might cause confusion why it doesn't work on static lists. In any case, if we ever want to override onclicks on static lists, we dont need to worry about backwards compatibility with this method.

**Comment by InuSasha** [comment]
@mkortstiege Is there a problem, why this don't merged into master? And how can i help to solve it?
The build failure was not result of this pr. ffmpeg build failed with seg-fault on windows :(
Or should i only wait?

After this pr is merged, i will pr the confluence change (https://github.com/InuSasha/xbmc/commit/dfcc9509470260014ab3e521bf9ab08ac7d128fe)

**Comment by Razzeee** [comment]
We're probably looking at this in the next days as we're on devcon right now and one thing we like to do is go through all PRs.

File: xbmc/GUIInfoManager.cpp [hunk 2]
```diff
@@ -2406,6 +2406,15 @@ bool CGUIInfoManager::EvaluateBool(const std::string &expression, int contextWin
   return result;
 }
 
+bool CGUIInfoManager::EvaluateBool(const std::string &expression, const CGUIListItemPtr &item)
```
Comment (mkortstiege) [inline xbmc/GUIInfoManager.cpp]: This is pretty much adding duped code. Mind refactoring it so it take context window and/or the listitem?

**Comment by mkortstiege** [comment]
Other than that i have no objections. It's a nice addition that make a lot of skinners happy. Thanks!

**Comment by mkortstiege** [comment]
@InuSasha care to handle the minor i pointed out. I really would like to merge it within this window.

**Comment by InuSasha** [comment]
@mkortstiege no problem, pr to my repo or merge it local and push together. 
I believe, you know what you are doing :)

**Comment by mkortstiege** [comment]
Oh, that's not what i meant ;) If you're low on time we can merge as is and i can handle the refactor later on if you like. Just let me know.

**Comment by InuSasha** [comment]
@mkortstiege sorry, bad school english.
if possible, i fix my mistakes by myself. so i learn it, and don't make it again :)
So: What have you pointed out.

**Comment by mkortstiege** [comment]
@InuSasha not a mistake but if possible I'd like to see `CGUIInfoManager::EvaluateBool` refactored to limit code duplication. Comment @ https://github.com/xbmc/xbmc/pull/8178#discussion_r41837322

**Comment by InuSasha** [comment]
@mkortstiege better?

**Comment by mkortstiege** [comment]
> @mkortstiege better?

That doesn't look right. You'll have to check for `item` before accessing it. Something like result = info->Get(item ? item.get() : 0 ); maybe?

**Comment by InuSasha** [comment]
no i have not to check. nullptr have get, and it return NULL :)
build and debug with gcc 4.9.2. works.

**Comment by mkortstiege** [comment]
Oh, well nevermind then.

**Comment by mkortstiege** [comment]
jenkins build this please.

**Comment by BigNoid** [comment]
You forgot to add the ```SetClickActions``` in ```GUIControlFactory.cpp``` for control types ```panel```, ```fixedlist``` and ```wraplist```. So atm it only works for control type ```list```.

**Comment by InuSasha** [comment]
@BigNoid added

**Comment by BigNoid** [comment]
Thanks. Working as intended now afaict.

**Comment by mkortstiege** [comment]
Thanks! jenkins build and merge.

**Pull request merged**